File: An 2/Medii de Proiectare si Programare/problema11final/src/main/java/project/problema11final/Service/Service.java
package project.problema11final.Service;

import project.problema11final.Domain.Entry;
import project.problema11final.Domain.Participant;
import project.problema11final.Domain.Race;
import project.problema11final.Domain.User;
import project.problema11final.Repository.EntryDB;
import project.problema11final.Repository.ParticipantDB;
import project.problema11final.Repository.RaceDB;
import project.problema11final.Repository.UserDB;
import project.problema11final.Service.DTOs.ParticipantDTO;
import project.problema11final.Service.DTOs.RaceDTO;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.UUID;

public class Service {
    private UserDB userDB;
    private RaceDB raceDB;
    private EntryDB entryDB;
    private ParticipantDB participantDB;

    public Service(UserDB userdb, RaceDB racedb, EntryDB entrydb, ParticipantDB participantdb) {
        this.userDB = userdb;
        this.raceDB = racedb;
        this.entryDB = entrydb;
        this.participantDB = participantdb;
    }

    public Iterable<User> getUsers() {
        return userDB.findAll();
    }

    public boolean LogIn(String username, String password) {
        Iterable<User> users = getUsers();
        for (User user : users) {
            if (user.getUsername().equals(username) && user.getPassword().equals(password)) {
                return true;
            }
        }
        return false;
    }

    public User getUserByUsername(String username){
        Iterable<User> users = getUsers();
        for(User user : users){
            if(user.getUsername().equals(username)){
                return user;
            }
        }
        return null;
    }

    public Iterable<Race> getRaces() {
        return raceDB.findAll();
    }

    public Iterable<RaceDTO> racesDTO(){
        Iterable<Race> races = raceDB.findAll();
        Iterable<Entry> entries = entryDB.findAll();
        HashSet<RaceDTO> raceDTOS = new HashSet<>();

        for (Race r : races){
            int noRacers = 0;
            for (Entry e : entries){
                if(e.getRaceId().equals(r.getId())){
                    noRacers += 1;
                }

                }
            RaceDTO raceDTO = new RaceDTO(String.valueOf(r.getEngineCapacity()), String.valueOf(noRacers));
            raceDTOS.add(raceDTO);
            }
        return raceDTOS;
        }

        public Iterable<String> getTeams(){
            Iterable<Participant> participants = participantDB.findAll();
            HashSet<String> teams = new HashSet<>();
            for (Participant p : participants){
                teams.add(p.getTeam());
            }
            return teams;
        }

        public Iterable<Entry> getEntries(){
            return entryDB.findAll();
        }

        public Iterable<Participant> getParticipants(){
            return participantDB.findAll();
        }

        public Iterable<ParticipantDTO> participantsDTO(){
            Iterable<Entry> entries = getEntries();
            Iterable<Participant> participants = getParticipants();

            ArrayList<ParticipantDTO> participantDTOS = new ArrayList<>();

//            for(Entry e : entries){
                for (Participant p : participants){
//                    if(e.getParticipantId().equals(p.getId())){
                        ParticipantDTO pDTO = new ParticipantDTO(p.getName(), p.getTeam(), String.valueOf(p.getEngineCapacity()));
                        participantDTOS.add(pDTO);
                    }
//                }
//            }

            return participantDTOS;
        }

        public Iterable<ParticipantDTO> getParticipantDTOforTeam(String team){
            Iterable<ParticipantDTO> participants = participantsDTO();
            ArrayList<ParticipantDTO> filtered = new ArrayList<>();
            for (ParticipantDTO p : participants){
                if(p.getTeam().equals(team)){
                    filtered.add(p);
                }
            }
            return filtered;
        }

    public void addParticipant(String username, String team, String engine) {
        Participant participant = new Participant(username, team, Integer.parseInt(engine));
        UUID id = UUID.randomUUID();
        participant.setId(id);
        participantDB.save(participant);
    }

    public Iterable<String> engineCapacities(){
        Iterable<Participant> participants = getParticipants();
        HashSet<String> engine = new HashSet<>();
        for (Participant p : participants){
            engine.add(String.valueOf(p.getEngineCapacity()));
        }
        return engine;
    }

    public void createEntry(Participant p){
        Iterable<Race> rs = getRaces();

        for (Race r : rs){
            if(p.getEngineCapacity() == r.getEngineCapacity()){
                Entry e = new Entry(r.getId(), p.getId());
                UUID id = UUID.randomUUID();
                e.setId(id);
                entryDB.save(e);

            }
        }
    }

    public void createAllEntries(){
        Iterable<Participant> ps = getParticipants();
        Iterable<Race> rs = getRaces();
        for (Participant p : ps ){
            for (Race r : rs) {
                if (p.getEngineCapacity() == r.getEngineCapacity()){
                    createEntry(p);
                }
            }
        }
    }

    public Participant getParticipant(String name, String team, String engineCapacity){
        for (Participant p : getParticipants()){
            if (p.getName().equals(name) && p.getTeam().equals(team) && String.valueOf(p.getEngineCapacity()).equals(engineCapacity)){
                return p;
            }
        }
        return null;
    }
}

File: An 2/Sisteme de Gestiunea Bazelor de Date/problema11ClientServer/Services/Service.cs
// using System;
// using System.Collections.Generic;
// using Model;
// using Persistence;
// using Networking.DTOs;
//
// namespace Services;
//
// public class Service
// {
//     private UserDB _userDb;
//     private RaceDB _raceDb;
//     private ParticipantDB _participantDb;
//     private EntryDB _entryDb;
//
//     public Service(UserDB userdb, RaceDB racedb, ParticipantDB participantdb, EntryDB entrydb)
//     {
//         this._userDb = userdb;
//         this._raceDb = racedb;
//         this._participantDb = participantdb;
//         this._entryDb = entrydb;
//     }
//
//     public IEnumerable<User> getUsers()
//     {
//         return _userDb.FindAll();
//     }
//     
//     public User Login(String username, String password)
//     {
//         IEnumerable<User> users = getUsers();
//         foreach (var user in users)
//         {
//             if (user.username.Equals(username) && user.password.Equals(password))
//             {
//                 return user;
//             }
//         }
//         return null;
//     }
//     
//     public IEnumerable<Race> getRaces()
//     {
//         return _raceDb.FindAll();
//     }
//
//     public IEnumerable<Entry> getEntries()
//     {
//         return _entryDb.FindAll();
//     }
//
//     public IEnumerable<RaceDTO> racesDTO()
//     {
//         IEnumerable<Race> races = getRaces();
//         IEnumerable<Entry> entries = getEntries();
//         List<RaceDTO> raceDtos = new List<RaceDTO>();
//         
//         foreach (var r in races)
//         {
//             int noRacers = 0;
//             foreach (var e in entries)
//             {
//                 if (e.raceId.Equals(r.Id))
//                 {
//                     noRacers++;
//                 }
//             }
//             RaceDTO raceDto = new RaceDTO(r.engineCapacity.ToString(), noRacers.ToString());
//             raceDtos.Add(raceDto);
//         }
//
//         return raceDtos;
//     }
//
//     public IEnumerable<Participant> getParticipants()
//     {
//         return _participantDb.FindAll();
//     }
//
//     public IEnumerable<string> getTeams()
//     {
//         IEnumerable<Participant> participants = getParticipants();
//         List<string> teams = new List<string>();
//         foreach (var p in participants)
//         {
//             if (!teams.Contains(p.team))
//             {
//                 teams.Add(p.team);
//             }
//         }
//         return teams;
//     }
//
//     public IEnumerable<ParticipantDTO> participantsDTO()
//     {
//         IEnumerable<Entry> entries = getEntries();
//         IEnumerable<Participant> participants = getParticipants();
//         
//         List<ParticipantDTO> participantDtos = new List<ParticipantDTO>();
//
//         foreach (var p in participants)
//         {
//             ParticipantDTO pDTO = new ParticipantDTO(p.name, p.team, p.engineCapacity.ToString());
//             participantDtos.Add(pDTO);
//         }
//
//         return participantDtos;
//     }
//
//     public IEnumerable<ParticipantDTO> getParticipantDTOforTeam(string team)
//     {
//         IEnumerable<ParticipantDTO> participantDtos = participantsDTO();
//         List<ParticipantDTO> teamParticipants = new List<ParticipantDTO>();
//         foreach (var p in participantDtos)
//         {
//             if (p.team.Equals(team))
//             {
//                 teamParticipants.Add(p);
//             }
//         }
//
//         return teamParticipants;
//     }
//
//     public void addParticipant(string username, string team, string engine)
//     {
//         Participant participant = new Participant(username, team, Int32.Parse(engine));
//         Guid id = Guid.NewGuid();
//         participant.Id = id;
//         _participantDb.save(participant);
//     }
//
//     public IEnumerable<string> engineCapacities()
//     {
//         IEnumerable<Participant> participants = getParticipants();
//         HashSet<string> engine = new HashSet<string>();
//         foreach (var p in participants)
//         {
//             engine.Add(p.engineCapacity.ToString());
//         }
//
//         return engine;
//     }
//
//     public IEnumerable<string> capacities()
//     {
//         IEnumerable<string> engines = engineCapacities();
//         List<string> capacities = new List<string>();
//         foreach (var e in engines)
//         {
//             capacities.Add(e);
//         }
//
//         return capacities;
//     }
//
//     public void createEntry(Participant p)
//     {
//         IEnumerable<Race> races = getRaces();
//
//         foreach (var r in races)
//         {
//             if(p.engineCapacity == r.engineCapacity)
//             {
//                 Entry entry = new Entry(r.Id, p.Id);
//                 Guid id = Guid.NewGuid();
//                 entry.Id = id;
//                 _entryDb.save(entry);
//             }
//         }
//     }
//     
//     public Participant getParticipant(string name, string team, string engine)
//     {
//         IEnumerable<Participant> participants = getParticipants();
//         foreach (var p in participants)
//         {
//             if (p.name.Equals(name) && p.team.Equals(team) && p.engineCapacity.ToString().Equals(engine))
//             {
//                 return p;
//             }
//         }
//
//         return null;
//     }
// }